### ISSUE TITLE
ansible_host: add to inventory without using private key files

### ISSUE THREAD

**Issue by sean-freeman** [ansible/terraform-provider-ansible#34]
## ansible_host: add to inventory without using private key files

TL;DR = The Terraform Provider for Ansible, is not using similar developer best practice measures as available with Terraform Provisioners.  See SSH Connection documentation > https://developer.hashicorp.com/terraform/language/resources/provisioners/connection

---

### Problem

Using this Terraform Provider for Ansible, still requires an SSH Private Key on the Terraform execution host's filesystem (e.g. in `/tmp`); the private key contents should be ephemeral to Terraform / in-memory during the Terraform execution for security purposes.

When defining the `ansible_host`,  it should have the correct args to generate an Ansible Inventory - similar to [ansible.builtin.add_host module](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/add_host_module.html)

See code snippets below for:
- Standard Ansible Inventory file
- Sample file of Problem
- Suggestion for Terraform Resource structure of ansible_host

---

### Standard Ansible Inventory file

```yaml
---

group1:
  hosts:
    host1:
      ansible_host: IP_GOES_HERE
      ansible_port: 22
      ansible_user: root
      ansible_ssh_private_key_file: "{{ playbook_dir }}/ssh/hosts_rsa"
      ansible_ssh_common_args: -o ConnectTimeout=180 -o ControlMaster=auto -o ControlPersist=3600s -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ForwardX11=no -o ProxyCommand='ssh -W %h:%p bastionuser@IP_GOES_HERE -p 50222 -i {{ playbook_dir }}/ssh/bastion_rsa -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null'
```

### Sample file of Problem

```hcl

### Create SSH Keys, load to AWS

# Create private SSH key only for Bastion/Jump Host usage
resource "tls_private_key" "bastion_ssh" {
  algorithm = "RSA"
}

# Create private SSH key for ssh connection of the host
resource "tls_private_key" "host_ssh" {
  algorithm = "RSA"
}

resource "aws_key_pair" "bastion_ssh" {
  key_name   = "test-ssh-keypair-bastion"
  public_key = trimspace(tls_private_key.bastion_ssh.public_key_openssh)
}

resource "aws_key_pair" "host_ssh" {
  key_name   = "test-ssh-keypair-host"
  public_key = trimspace(tls_private_key.host_ssh.public_key_openssh)
}


### Create AWS EC2 virtual server, use AWS Key Pair containing the SSH Public Key

resource "aws_instance" "bastion" {
# add many args here
}

resource "aws_instance" "host" {
# add many args here
}


### Download Ansible Playbook to Terraform execution directory

resource "null_resource" "ansible_playbooks_repo_download" {
  provisioner "local-exec" {
    command = <<EOT
    curl -L https://github.com/gh-org/repo/archive/refs/heads/main.tar.gz -o ${path.root}/tmp/ansible_playbooks_repo-main.tar.gz
    mkdir -p ${path.root}/tmp/ansible_playbooks_repo
    tar xvf ${path.root}/tmp/ansible_playbooks_repo-main.tar.gz --strip-components=1 -C ${path.root}/tmp/ansible_playbooks_repo
    EOT
  }
}


### Create Ansible Inventory

resource "ansible_group" "group1" {
  name     = "group1"
}

resource "ansible_host" "host" {
  name      = var.hostname
  groups    = ["group1"]

  variables = {
    ansible_user                 = "root"
    ansible_host                 = aws_instance.host.private_ip
    ansible_ssh_private_key_file = var.ssh_host_private_key_path
    ansible_python_interpreter   = "/usr/bin/python3"
    ansible_ssh_common_args      =  yamldecode(<<-EOT
                                      string_input: >-
                                      -o ConnectTimeout=180 -o ControlMaster=auto
                                      -o ControlPersist=3600s -o StrictHostKeyChecking=no
                                      -o UserKnownHostsFile=/dev/null -o ForwardX11=no
                                      -o ProxyCommand='ssh -W %h:%p ${var.bastion_os_user}@${var.bastion_public_ip}
                                      -p ${var.bastion_port} -i ${var.ssh_bastion_private_key_path} -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
                                    EOT
                                    ).string_input
  }
}


### Run Ansible Playbook using the created Ansible Inventory

resource "ansible_playbook" "playbook" {
  playbook   = "${path.root}/tmp/ansible_playbooks_repo/playbook.yml"
  name       = "localhost"
  replayable = false

  extra_vars = {
    ansible_connection = "local"
    var_a              = "Some variable"
    var_b              = "Another variable"
  }
}
```

### Suggestion for Terraform Resource structure of ansible_host

```hcl
resource "ansible_host" "host" {
  name                = var.hostname
  inventory_groups    = ["group1"]

  # Follow syntax of Terraform Prosivioners, generate the ansible_ssh_common_args
  connection {
    user                = "root"                                       ## ansible_user
    host                = aws_instance.host.private_ip                 ## ansible_host
    private_key         = tls_private_key.host_ssh.private_key_pem     ## ansible_ssh_private_key_file
    bastion_host        = aws_instance.bastion.public_ip               ## generate string for ansible_ssh_common_args
    bastion_port        = 22                                           ## generate string for ansible_ssh_common_args
    bastion_user        = "ec2-user"                                   ## generate string for ansible_ssh_common_args
    bastion_private_key = tls_private_key.bastion_ssh.private_key_pem  ## generate string for ansible_ssh_common_args
  }
}
```

**Comment by anazobec** [comment]
Hi, in Ansible it is possible to define almost any Ansible setting using [(magic) variables](https://docs.ansible.com/ansible/latest/reference_appendices/special_variables.html), which is why we chose to use this 1-to-1 mapping implementation and why ``variables`` (in ``ansible_host`` and ``ansible_group`` resources) is named ``variables``, since you don't necessarily have to pass in/define only Ansible connection settings. With this explanation aside, it wouldn't be possible to rename ``variables`` to ``connection`` since the word ``connection`` is a [reserved word used by Terraform provisioners](https://developer.hashicorp.com/terraform/language/resources/provisioners/connection) and ``ansible_host`` and ``ansible_group`` are not provisioners but resources. At most, we could rename it to something like ``connection_settings`` or ``connection_setup`` instead.

**Comment by sean-freeman** [comment]
Hi @anazobec

I understand the principle behind `variables` implementation for both Terraform Resources `ansible_host` and `ansible_group`. This is a flexible approach that allows the 1-to-1 mapping, that makes sense.

However, I would encourage further design research for _(`connection_settings` or `connection_setup` as you suggested)_ dynamically generating the Ansible connection settings.

To be verbose about the intent behind this GH Issue....

> `Terraform Resource "null_resource" with provisioner "remote-exec".... connection:`
- Target Host
    - OS User
    - IP
    - Private Key PEM as string...  [NOT using Private Key PEM _**`file path`**_ as string]
- Bastion Host
    - OS User
    - IP
    - Private Key PEM as string...  [NOT using Private Key PEM _**`file path`**_ as string]

> `Terraform Resource ansible_host or ansible_group.... variables:`
- Target Host
    - OS User
    - IP
    - Private Key _**`file path`**_ as string  [UNSECURE, requires PEM file on Terraform execution host]
- Bastion Host
    - OS User
    - IP
    - Private Key _**`file path`**_ as string  [UNSECURE, requires PEM file on Terraform execution host]


The private key contents should be ephemeral / in-memory during the Terraform execution for security purposes. The user should not have to output the SSH Private Key PEM file contents to the Terraform execution host, before any Ansible Inventory can be built for an Ansible Playbook execution.

**Comment by sean-freeman** [comment]
@anazobec any further update please?

**Comment by 65156** [comment]
seems like there is an intention to ensure this is not secure.

**Comment by sean-freeman** [comment]
@65156 It has been ignored, and it breaks all potential usage of this Terraform Provider with any 'runner' implementations (e.g. Terraform Cloud/Enterprise, Spacelift, Env0, Azure DevOps Pipelines, IBM Cloud Schematics etc etc) for calling Terraform > Ansible. Only @mandar242 is committing code for the past 12 months.

**Comment by jillr** [comment]
Hi @sean-freeman .  Thank you for the detailed information about the use case in this ticket, the examples are super helpful.  You're right, running this provider in another pipeline is not a deployment scenario that is currently optimized for in this provider. 

We are continuing to expand our support of Terraform based workflows with Ansible, so while contributions in this particular repo have been slower in the last few months we are still investing in this area overall.  @tima is at configmgmt camp this week (where he presented on orchestrating Terraform with Ansible earlier today https://cfp.cfgmgmtcamp.org/2024/talk/KQHVVK/), I'd like him to chime in on this one when he gets back.

**Comment by marshallford** [comment]
I've been following this issue for a while (since my initial evaluation of the provider), and it's one I keep revisiting without a satisfying resolution. The core challenge, as I see it, is how to avoid persisting sensitive data on the Terraform execution host, given that the integration point with Ansible relies (more or less) on `exec.Command`.

These are the angles I’ve been exploring so far:

## Constraints

- **Ansible CLI behavior:** As mentioned above, Ansible primarily accepts input via CLI args and files. While it's theoretically possible to use the Python `ansible-runner` library or another in-memory interface from the Golang, that approach seems impractical for now.

- **Containerization:** Running Terraform or Ansible inside a container doesn’t eliminate the issue -- it simply relocates it. Sensitive data still ends up written to disk (e.g., `/var/lib/docker/overlay2`), just in a different location.

- **In-memory volumes are complex and platform-dependent:** Using in-memory file systems introduces its own set of tradeoffs (e.g., swap behavior, size limits). More importantly, there’s no reliable cross-platform abstraction for this in Go -- especially for Windows targets, where options appear limited or nonexistent.

## Interim Mitigations

- **Provider inputs:** Instead of relying on `local-exec` or external pipeline steps to write files that Ansible can consume, the provider could take responsibility for managing these inputs internally.

- **Cleanup:** Ensure any temporary files created for Ansible are deleted immediately after execution.

- **Configurable temp path:** Allow users to configure the temporary directory, giving them more control. For example, Linux users could point temp files to a tmpfs mount like `/dev/shm`, if they’re comfortable with the associated tradeoffs.
